feat: Add render_to_formspec_string function

Lazerbeak12345 opened the pull request:
Adds `Form:render_to_formspec_string` function.

Abstracted:

- `Form:show` to seperate out the next_formspec logic into the new private `new_next_formname` function.
- `on_fs_input` to seperate out the callback logic into the new private `fs_process_events` function.

`Form:render_to_formspec_string` is a new public function, taking a player ref and a context. Makes use of `new_next_formname` to get a unique formname for debugging purposes. Returns a tuple of the formspec and a asdfs function, which accepts a `fields` reference. This function calls `fs_process_events` to trigger the form events, and returns a boolean indicating if the form should be redrawn.

see #3 

## TODO (before I mark ready for review)

- [x] Code
- [x] Unit tests
  - [x] standalone
  - [x] embedded
- [x] Review comments and concerns
- [x] Documentation

Lazerbeak12345 committed:
feat: Add render_to_formspec_string function

Abstracted:

- `Form:show` to seperate out the next_formspec logic into the new
  private `new_next_formname` function.
- `on_fs_input` to seperate out the callback logic into the new
  private `fs_process_events` function.

`Form:render_to_formspec_string` is a new public function, taking a
player ref and a context. Makes use of `new_next_formname` to get a
unique formname for debugging purposes. Returns a tuple of the
formspec and a asdfs function, which accepts a `fields` reference. This
function calls `fs_process_events` to trigger the form events, and
returns a boolean indicating if the form should be redrawn.

Lazerbeak12345 committed:
refactor(lint): Line was too long for linter

Lazerbeak12345 committed:
test(render-to-formspec): test render-to-formspec

luk3yx commented on test.lua:
```diff
@@ -314,4 +314,51 @@ describe("Flow", function()
+            assert.equals(manual_spy_count, 1, "event passsed down only once")
+            assert.equals(manual_spy[1], player, "player was first arg")
+            assert.equals(manual_spy[2], ctx, "context was next")
+            assert.equals(manual_spy[3], fields.btn, "the value was last")
```
The third parameter has always been undocumented (I probably added it before I thought of storing values in `ctx`) and I may eventually remove it, testing for it probably isn't a good idea (though I guess I can change the tests if/when I remove it).

Lazerbeak12345 replied:
fixed in [a7744d5](https://github.com/luk3yx/minetest-flow/pull/5/commits/a7744d5357a7928f940fdfae7d28fe851d55cb88)

luk3yx commented on init.lua:
```diff
@@ -820,6 +824,34 @@ function Form:set_as_inventory_for(player, ctx)
+
+-- Returns a tuple of string and function
+function Form:render_to_formspec_string(player, ctx)
+    local formname = ("flow:subform_%x"):format(next_formname)
```
Does this need a formname?

Lazerbeak12345 replied:
fixed in 85f68de

luk3yx commented on init.lua:
```diff
@@ -820,6 +824,34 @@ function Form:set_as_inventory_for(player, ctx)
+-- Declared here to be accessable by render_to_formspec_string
+local fs_process_events
+
+-- Returns a tuple of string and function
```
What about `size[]`?

Lazerbeak12345 replied:
Isn't this size information provided by the root element returned from the callback to `flow.make_gui`?

luk3yx replied:
The `size[]` is returned in the formspec but if you want to embed it in an inventory mod then it would need to be removed.

Lazerbeak12345 replied:
Is this true for `formspec_version[]` as well?

luk3yx replied:
Yes. I'm not sure what the best way to return them would be

Lazerbeak12345 replied:
From testing, this doesn't seem to actually cause any problems to have these, but if we'd like to stay on spec, conditionally removing `size[]` looks easy (can be done the same way `no_prepend[]` is done), and after render, `formspec_version[]` can be removed with RegExps.

Should we even worry about this, since it works without removing them? Is there a better way to remove `formspec_version[]`?

luk3yx replied:
> Should we even worry about this, since it works without removing them?

I'd rather they be returned separately, I'd rather not rely on MT not erroring if there is `size[]` in the middle of formspecs.

> Is there a better way to remove formspec_version[]?

Maybe something like this?

```lua
local tree, form_info = form:_render(...)
local size = table.remove(tree, 1)
tree.formspec_version = nil
local fs = assert(formspec_ast.unparse(tree))

return fs, function(...)
    -- Handle fields
end, size.x, size.y
```

Lazerbeak12345 replied:
I figured I'd do this via an optional boolean as the third arg. Some apis, like sfinv, allow you to just set the whole formspec (`sfinv.register_page`) and some only embed, so both are needed.

luk3yx replied:
> I figured I'd do this via an optional boolean as the third arg. Some apis, like sfinv, allow you to just set the whole formspec (`sfinv.register_page`) and some only embed, so both are needed.

`sfinv.make_formspec` has a separate argument for the size element:

```lua
sfinv.register_page("test:page", {
    title = "Test page",
    get = function(self, player, context)
        local fs, on_event, w, h = form:render_to_formspec_string(...)
        context.test_page_event = on_event
        return sfinv.make_formspec(player, context, fs, true, ("formspec_version[6]size[%s,%s]"):format(w, h))
    end,
    on_player_receive_fields = function(self, player, context, fields)
        if not context.test_page_event or context.test_page_event(fields) then
            sfinv.set_player_inventory_formspec(player, context)
        end
    end,
})
```

Lazerbeak12345 replied:
Right. Thought I'd clear a miscommunication with how I actually implemented it:

```lua
sfinv.register_page("test:page", {
    title = "Test page",
    get = function(self, player, context)
        -- Info is only present when the third arg is truthy (embed mode)
        local fs, on_event, info = form:render_to_formspec_string(..., true)
        context.test_page_event = on_event
        return sfinv.make_formspec(player, context, fs, true, ("formspec_version[%s]size[%s,%s]"):format(info.formspec_version, info.w, info h))
    end,
    on_player_receive_fields = function(self, player, context, fields)
        if not context.test_page_event or context.test_page_event(fields) then
            sfinv.set_player_inventory_formspec(player, context)
        end
    end,
})
```

It's an argument to `Flow:render_to_formspec_string` I was talking about. I was trying to point out that sometimes we _do_ want the size and version, but other times we don't, so it should be an arg.

luk3yx replied:
I wasn't sure where `render_to_formspec_string` could be used without setting `embedded`, I guess it could be used if you want to create a sfinv page without the navigation bar.

Lazerbeak12345 replied:
Sorry to reopen. It occurs to me that there's a possibility we want embedded mode to be the default. What do you think?

luk3yx replied:
That sounds like a good idea.

luk3yx reviewed (commented):
I'm not as sure about adding this feature now that `set_as_inventory_for` exists, though if you were planning on using it then it's probably okay.

I want to strongly discourage people from using flow forms with node meta somehow as well, though maybe a warning in the documentation is enough to do this.

luk3yx commented on init.lua:
```diff
@@ -820,6 +824,34 @@ function Form:set_as_inventory_for(player, ctx)
+function Form:render_to_formspec_string(player, ctx)
+    local formname = ("flow:subform_%x"):format(next_formname)
+    new_next_formname()
+    local fs, form_info = prepare_form(self, player, formname, ctx or {})
```
I think `auto_name_id` should be preserved somehow

Lazerbeak12345 replied:
I'm not sure how I might go about this. Considering that the automatic names _do_ need to be unique, this actually does cause a problem if a parent formspec contains two seperate flow formspecs.

Should this be an argument to `Form:render_to_formspec_string`, do you think we should track it akin to how we do `next_formname` or do you have something else in mind?

luk3yx replied:
~~I don't like the idea of using a global counter since `auto_name_id` wraps around at a much lower number, but I'm also not sure about adding per-player counters.~~

I think that having a table that stores per-player values (to prevent another player from trying to make the counter wrap around) would be the best way of doing it, though I'm open to other suggestions.

Lazerbeak12345 replied:
Yeah sure! I like that solution. I think that providing `auto_name_id` as an optional third argument would be wise as well, so it can be overridden by the API user.

Lazerbeak12345 replied:
Not sure if that third argument is needed. I can't think of a usecase for it where it would actually be reliable anyway.

Lazerbeak12345 replied:
What do you think about the latest commit?

luk3yx replied:
I think it's okay, though table entries should be removed when players leave the game.

Lazerbeak12345 replied:
Done, and pushed. I'll resolve this convo, but feel free to reopen if you've got further concerns.

Lazerbeak12345 commented:
> I'm not as sure about adding this feature now that `set_as_inventory_for` exists, though if you were planning on using it then it's probably okay.

My idea was that this could be used to have new mods make use of things like sfinv where a formspec string is expected, but then be able to use `flow` for newer mods, like my experiment.

> I want to strongly discourage people from using flow forms with node meta somehow as well, though maybe a warning in the documentation is enough to do this.

Yeah, I'd recommend a warning like that. I'm not sure I understand why you're not recommending node meta, though.

Lazerbeak12345 committed:
test(render-to-formspec): don't test last arg

The last arg to the event callback isn't documented.

Lazerbeak12345 committed:
refactor(next_formname): doesn't need to be general

luk3yx commented:
> Yeah, I'd recommend a warning like that. I'm not sure I understand why you're not recommending node meta, though.

Node meta formspecs are the same for every player that opens them, which would break `ctx.form` (other players with the node formspec open would get form values typed in by another player if the form is redrawn) and would mean the formspec version would be unknown.

Lazerbeak12345 commented:
> Node meta formspecs are the same for every player that opens them, which would break `ctx.form` (other players with the node formspec open would get form values typed in by another player if the form is redrawn) and would mean the formspec version would be unknown.

This makes sense. In documentation then, we should warn something like:

> Use of Flow in Node metadata isn't supported. All players would get the same values typed in by other players, and flow would be unable to generate formspecs targeted to specific versions that these players may need.

It should be noted that `render_to_formspec_string` _does_ require a player object, since it's passed down through the usual render pipeline. They'd have to do a workaround to get it to work in node metadata in the first place. It can also be called multiple times without conflict, as this function is stateless.

Such a workaround could be:

- On server start define the form.
- Keep track of the lowest formspec version needed. Every time the lowest formspec changes, regenerate the nodemeta.
- Keep a table of the event functions returned by `render_to_formspec_string` per player.
- When the actual callback happens, call the proper event function only.
- Intercept a `quit` event, and re-render the form with fresh context.

This doesn't fix the globally visible input strings, but since that's already a problem in minetest already, I've noticed that node meta inputs are almost always global anyways (like signs or luacontroller) or simply re-render with a fresh context on a form submission.

Lazerbeak12345 commented:
> Such a workaround could be:
> 
> * On server start define the form.
> * Keep track of the lowest formspec version needed. Every time the lowest formspec changes, regenerate the nodemeta.
> * Keep a table of the event functions returned by `render_to_formspec_string` per player.
> * When the actual callback happens, call the proper event function only.
> * Intercept a `quit` event, and re-render the form with fresh context.

This actually could run into problems, since the event function expects the _exact same_ form names, and `auto_name_id` would not be sufficient. They'd have to manually name everything. This might also need to be added to the warning? Or to be a warning of its own.

Lazerbeak12345 committed:
Merge branch 'main' into 2-render-to-formspec

Lazerbeak12345 commented:
Resolving merge conflicts to stay up-to-date enough with main.

Lazerbeak12345 committed:
fix(typo): Merge conflict had typo

Lazerbeak12345 committed:
test(render-to-formspec): getbyname should assert

Lazerbeak12345 committed:
fix(render-to-formspec): retain auto_name_id

Keeps the auto_name_id number in a table, indexed by player name.

See comments.

Lazerbeak12345 committed:
fix: Automaticly remove auto_name_ids

Done on leaveplayer, applies to `render_to_formspec_auto_name_ids`

luk3yx commented on init.lua:
```diff
@@ -894,12 +932,8 @@ function flow.make_gui(build_func)
-    local form_info = form_infos[name]
-    if not form_info or formname ~= form_info.formname then return end
-
+-- Declared locally above to be accessable to render_to_formspec_string
```
```suggestion
-- Declared locally above to be accessible to render_to_formspec_string
```

Lazerbeak12345 committed:
docs: fix typo in init.lua comment

Co-authored-by: luk3yx <luk3yx@users.noreply.github.com>

Lazerbeak12345 committed:
feat(render-to-formspec): Add proper embed support

luk3yx committed:
Code style changes

luk3yx committed:
Merge branch 'main' into pr

luk3yx committed:
Fix typo

luk3yx committed:
Attempt to write documentation

luk3yx committed:
Put experimental features into <details>

luk3yx committed:
Update README.md

luk3yx commented:
Is the documentation I've written okay?

luk3yx committed:
Update README.md

Lazerbeak12345 commented:
I think so, pending API change discussion ofc

Lazerbeak12345 commented on README.md:
```diff
@@ -434,3 +436,26 @@ example_inventory:unset_as_inventory_for(player)
+ - The `process_event(fields)` callback will return true if the formspec should
+   be redrawn, where `render_to_formspec_string` should be called and the new
+   `process_event` should be used in the future. This function may return true
+   even if fields.quit is sent.
```
Should this behave like this? What would it return instead, under what conditions?

luk3yx replied:
I think it's fine since the formspec may need to be updated if it's going to get shown again (for example in sfinv).

Lazerbeak12345 committed:
refactor(embed mode)!: Invert embed mode

BREAKING CHANGE: Inverts the embeded argument to
`Form:render-to-formspec`. This is a more sensible default, considering
the usecases in mind. Setting as the root formspec should be _very_
rare.

luk3yx commented on init.lua:
```diff
@@ -842,6 +842,51 @@ function Form:set_as_inventory_for(player, ctx)
+-- Unique per-user to prevent players from making the counter wrap around for
+-- other players.
+local render_to_formspec_auto_name_ids = {}
+-- if `embedded` is true, then this won't output a "root" formspec, but instead
+--   will output  formspec designed to be embedded along with the information
+--   that was removed from the form
+function Form:render_to_formspec_string(player, ctx, embedded)
+    local name = player:get_player_name()
+    local info = minetest.get_player_information(name)
+    local tree, form_info = self:_render(player, ctx or {},
+        info and info.formspec_version, render_to_formspec_auto_name_ids[name])
+    local public_form_info
+    if embedded then
+        local size = table.remove(tree, 1)
+        public_form_info = {w = size.w, h = size.h,
+            formspec_version = tree.formspec_version}
+        tree.formspec_version = nil
+    end
```
```suggestion
-- If standalone is set, this will return a standalone formspec, otherwise it
-- will return a formspec that can be embedded and a table with its size and
-- target formspec version
function Form:render_to_formspec_string(player, ctx, standalone)
    local name = player:get_player_name()
    local info = minetest.get_player_information(name)
    local tree, form_info = self:_render(player, ctx or {},
        info and info.formspec_version, render_to_formspec_auto_name_ids[name])
    local public_form_info
    if not standalone then
        local size = table.remove(tree, 1)
        public_form_info = {w = size.w, h = size.h,
            formspec_version = tree.formspec_version}
        tree.formspec_version = nil
    end
```

Lazerbeak12345 replied:
sorry, hadn't seen this review

luk3yx commented on init.lua:
```diff
@@ -850,16 +850,16 @@ local fs_process_events
         info and info.formspec_version, render_to_formspec_auto_name_ids[name])
     local public_form_info
-    if embedded then
+    if standalone then
```
```suggestion
    if not standalone then
```

luk3yx commented on init.lua:
```diff
@@ -850,16 +850,16 @@ local fs_process_events
---   will output  formspec designed to be embedded along with the information
---   that was removed from the form
-function Form:render_to_formspec_string(player, ctx, embedded)
+-- if `standalone` is not true, then this won't output a "root" formspec, but
+--   instead will output  formspec designed to be embedded along with the
+--   information that was removed from the form
```
```suggestion
-- If standalone is set, this will return a standalone formspec, otherwise it
-- will return a formspec that can be embedded and a table with its size and
-- target formspec version
```

Lazerbeak12345 replied:
Just applied suggestion.

luk3yx commented on init.lua:
```diff
@@ -842,6 +842,51 @@ function Form:set_as_inventory_for(player, ctx)
+        end
+        return fs_process_events(player, form_info, fields)
+    end
+    if standalone then
+        return fs, event, public_form_info
+    else
+        return fs, event
+    end
```
The if/else statement could be removed entirely, though then it'd return `fs, event, nil` when standalone is set.

```suggestion
    return fs, event, public_form_info
```

Lazerbeak12345 replied:
Added in latest commit.

Lazerbeak12345 committed:
fix(standalone)!: Booleans weren't correct

Also included better comment wording from review suggestion.

BREAKING CHANGE: I forgot to invert the booleans in the last commit

Co-authored-by: luk3yx <luk3yx@users.noreply.github.com>

Lazerbeak12345 committed:
docs(comments): Further suggested grammar changes

Co-authored-by: luk3yx <luk3yx@users.noreply.github.com>

Lazerbeak12345 committed:
feat: Always return a value

Old one was a bit harder to understand

Co-authored-by: luk3yx <luk3yx@users.noreply.github.com>

Lazerbeak12345 commented:
We've now got 3 failing unit tests. I'll see if I can fix them before I add the new test.

Lazerbeak12345 commented:
two of them were due to fa13e18.

luk3yx commented:
> two of them were due to [fa13e18](https://github.com/luk3yx/minetest-flow/commit/fa13e188f97d0e5ba62a288c84168eb0f9e69004).

Oops, I forgot to update the formspec version in `normalise_tree()` in test.lua. Maybe I need to figure out how to make the unit tests run in a GitHub workflow

Lazerbeak12345 commented:
I also realized I did `assert(a == b)` when I should have done `assert.are.equal(a, b)` in quite a few places. I might make that another PR later, but it's a pretty quick fix. (Benefit is error readability)

luk3yx committed:
The tests seem to pass for me without this

luk3yx committed:
Actually fix unit tests

I was running busted on an older version of the code before

Lazerbeak12345 committed:
test(standalone): new standalone test

Ensure we test both cases

luk3yx committed:
Fix test

luk3yx commented:
Thanks
